### PR TITLE
[PIPELINES 4] Create release based on annotated git tag

### PR DESCRIPTION
**Pull request by petrkotas**
### Which issue this PR addresses:

Fixes https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9739113/
Fixes https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9739185/
Fixes https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9741850/

### What this PR does / why we need it:

Changes to makefile:
- Introduces two new entrypoints in the `Makefile` that build and push tagged container image.
- Get tag only when it is annotated and attached to current commit, otherwise error.

Changed to pipelines:
- Add second build pipeline, 
	- chained after successful E2E
	- triggered when tag started the pipelines
	- Generates changelog from commits and summary provided in tag annotation
- Add second build image template 
	- build tagged image
	- needs to be added as a pipelines into the AZP

Signed-off-by: Petr Kotas <pkotas@redhat.com>

### Test plan for issue:

Manual. 

### Is there any documentation that needs to be updated for this PR?

Doc page once all bits are merged. The whole pipeline will get diagram and dedicated documentation page.

### CODE REVIEW

**Comment by petrkotas** [comment]
Contains commit from a https://github.com/Azure/ARO-RP/pull/1674 which need to go first.

File: .pipelines/build-and-push-images.yml
Comment (petrkotas) [inline .pipelines/build-and-push-images.yml:16]: Unfortunately AZP does not support better filtering. No regex at this point.

File: .pipelines/build-and-push-images.yml
Comment (petrkotas) [inline .pipelines/build-and-push-images.yml:11]: trigger after successful E2E run.

File: .pipelines/build-and-push-images.yml
Comment (petrkotas) [inline .pipelines/build-and-push-images.yml:43]: Extracts the annotation from the annotated tag. This is later used as input text for summary of the changelog.
The text have to be stored as a file as azp task can only load file.

File: .pipelines/build-and-push-images.yml
Comment (petrkotas) [inline .pipelines/build-and-push-images.yml:45]: Release task generating the automated changelog in Github/Releases. It require service connection of type Github to Azure/ARO-RP granted by access token created at https://github.com/settings/tokens with write permission to the ARO-RP repository.

File: .pipelines/ci.yml
Comment (petrkotas) [inline .pipelines/ci.yml:6]: Trigger CI on any tag. Unfortunately no filtering available here.

File: .pipelines/deploy-azure-env.yml
Comment (petrkotas) [inline .pipelines/deploy-azure-env.yml:10]: Part of https://github.com/Azure/ARO-RP/pull/1674

File: .pipelines/int-release.yml
Comment (petrkotas) [inline .pipelines/int-release.yml:13]: Part of https://github.com/Azure/ARO-RP/pull/1674

File: .pipelines/prod-release.yml
Comment (petrkotas) [inline .pipelines/prod-release.yml:9]: Part of https://github.com/Azure/ARO-RP/pull/1674

File: .pipelines/templates/template-job-deploy-azure-env.yml
Comment (petrkotas) [inline .pipelines/templates/template-job-deploy-azure-env.yml:31]: Part of https://github.com/Azure/ARO-RP/pull/1674

File: .pipelines/templates/template-push-images-to-acr.yml
Comment (petrkotas) [inline .pipelines/templates/template-push-images-to-acr.yml:15]: At this point this is direct copy of the previous step. First the steps are duplicated, once all works old can be deleted.

File: Makefile
Comment (petrkotas) [inline Makefile:2]: TAG is parsed by `git describe`, which returns nothing if there is no annotated tag.

**Comment by petrkotas** [comment]
E2E failed on tag push. Please do not rerun E2E. Build image got run because the trigger phase is not yet present in the pipelines and the code from this PR is ignored. Creating separate PR for trigger.

**Comment by petrkotas** [comment]
Converting to draft as I have figured out the path forward, which will require slight changes.

File: .pipelines/ci.yml
Comment (petrkotas) [inline .pipelines/ci.yml:9]: Added in https://github.com/Azure/ARO-RP/pull/1685

File: .pipelines/build-and-push-images-tagged.yml
Comment (petrkotas) [inline .pipelines/build-and-push-images-tagged.yml:1]: Newly added pipeline chained after successful E2E pipeline when tag is the trigger. 

Requires adding as new pipeline to the AZP.

Duplicates the original pipeline, however it produces tagged image release. It also creates github release, which requires GitHub service connection.

File: .pipelines/build-and-push-images-tagged.yml
Comment (petrkotas) [inline .pipelines/build-and-push-images-tagged.yml:43]: Extracts release summary from annotated tag. The summary is stored in the file, which is used as `releaseNotesFile` input for `GithubRelease@0` tasks. Task does not support other form of input.

Task also require `githubConnection`, which is service connection in ADO of type Github pointing to `ARO-RP` repository with write access.

Task also generates auto-changelog based on commit titles. To make it work nicely [conventional commits](https://conventionalcommits.org/en/v1.0.0/) are required.

File: .pipelines/deploy-azure-env.yml
Comment (petrkotas) [inline .pipelines/deploy-azure-env.yml:10]: Part of https://github.com/Azure/ARO-RP/pull/1674 which needs to go first.

File: .pipelines/int-release.yml
Comment (petrkotas) [inline .pipelines/int-release.yml:13]: Part of https://github.com/Azure/ARO-RP/pull/1674 which needs to go first.

File: .pipelines/prod-release.yml
Comment (petrkotas) [inline .pipelines/prod-release.yml:9]: Part of https://github.com/Azure/ARO-RP/pull/1674 which needs to go first.

File: .pipelines/templates/template-job-deploy-azure-env.yml
Comment (petrkotas) [inline .pipelines/templates/template-job-deploy-azure-env.yml:31]: Part of https://github.com/Azure/ARO-RP/pull/1674 which needs to go first.

File: .pipelines/templates/template-push-images-to-acr-tagged.yml
Comment (petrkotas) [inline .pipelines/templates/template-push-images-to-acr-tagged.yml:13]: Copies the original pipeline, this is the only difference. Call tagged makefile entrypoint.

File: .pipelines/build-and-push-images-tagged.yml
Comment (petrkotas) [inline .pipelines/build-and-push-images-tagged.yml:28]: Uses duplicated template with change to call tagged makefile entrypoint.

File: Makefile
Comment (petrkotas) [inline Makefile:2]: Read the annotated tag from git. Requires the tag to be at current commit. Error otherwise.

File: Makefile
Comment (petrkotas) [inline Makefile:56]: When no tag is present end. This happens when tag is not on current commit.

File: .pipelines/templates/template-push-images-to-acr-tagged.yml
Comment (petrkotas) [inline .pipelines/templates/template-push-images-to-acr-tagged.yml:10]: `TAG` is set here for visibility, it is also set in the `Makefile` for local work.

File: Makefile
Comment (petrkotas) [inline Makefile:2]: Set variable when not already set

**Review by bryanro92 — APPROVED** [review]

**Review by bennerv — APPROVED** [review]
LGTM - wait for previous PRs to merge before this one.

**Review by s-amann — APPROVED** [review]

**Pull request merged**
